orchestrator/relayer/batch_relaying.go

```diff
@@ -2,88 +2,247 @@ package relayer
-		Msg("found latest valsets")
+	ethBlockHeight := lastEthereumHeader.Number.Uint64()
 
-	if oldestSignedBatch.BatchNonce > latestEthereumBatch.Uint64() {
+	for tokenContract, batches := range possibleBatches {
 
+		// requests data from Ethereum only once per token type, this is valid because we are
+		// iterating from oldest to newest, so submitting a batch earlier in the loop won't
+		// ever invalidate submitting a batch later in the loop. Another relayer could always
+		// do that though.
 		latestEthereumBatch, err := s.peggyContract.GetTxBatchNonce(
 			ctx,
-			common.HexToAddress(oldestSignedBatch.TokenContract),
+			tokenContract,
 			s.peggyContract.FromAddress(),
 		)
 		if err != nil {
+			s.logger.Err(err).Msg("failed to get latest Ethereum batch")
 			return err
 		}
-		// Check if oldestSignedBatch already submitted by other validators in mean time
-		if oldestSignedBatch.BatchNonce > latestEthereumBatch.Uint64() {
+
+		// now we iterate through batches per token type
+		for _, batch := range batches {
+
+			if batch.Batch.BatchTimeout < ethBlockHeight {
+				s.logger.Debug().
+					Uint64("batch_nonce", batch.Batch.BatchNonce).
+					Str("token_contract", batch.Batch.TokenContract).
+					Uint64("batch_timeout", batch.Batch.BatchTimeout).
+					Uint64("eth_block_height", ethBlockHeight).
+					Msg("batch has timed out and can't be submitted")
+				continue
+			}
+
+			// if the batch is newer than the latest Ethereum batch, we can submit it
+			if batch.Batch.BatchNonce <= latestEthereumBatch.Uint64() {
+				continue
+			}
+
+			txData, err := s.peggyContract.EncodeTransactionBatch(ctx, currentValset, batch.Batch, batch.Signatures)
+			if err != nil {
+				return err
+			}
+
+			estimatedGasCost, gasPrice, err := s.peggyContract.EstimateGas(ctx, s.peggyContract.Address(), txData)
+			if err != nil {
+				s.logger.Err(err).Msg("failed to estimate gas cost")
+				return err
+			}
+
+			// TODO: estimate gas cost and check if this tx is profitable
+			// If the batch is not profitable, move on to the next one.
+			if !s.IsBatchProfitable(ctx, batch.Batch, estimatedGasCost, gasPrice) {
+				continue
+			}
+
 			s.logger.Info().
-				Uint64("latest_batch", oldestSignedBatch.BatchNonce).
+				Uint64("latest_batch", batch.Batch.BatchNonce).
 				Uint64("latest_ethereum_batch", latestEthereumBatch.Uint64()).
-				Msg("we have detected latest batch but Ethereum has a different one. Sending an update!")
+				Msg("we have detected a newer profitable batch. Sending an update!")
 
-			// Send SendTransactionBatch to Ethereum
-			txHash, err := s.peggyContract.SendTransactionBatch(ctx, currentValset, oldestSignedBatch, oldestSigs)
+			_, err = s.peggyContract.SendTransactionBatch(ctx, txData)
 			if err != nil {
 				return err
 			}
-			s.logger.Info().Str("tx_hash", txHash.Hex()).Msg("sent Ethereum Tx (TransactionBatch)")
+
+			// update our local tracker of the latest batch
+			s.lastSentBatchNonce = batch.Batch.BatchNonce
```

Comment:
There aren't multiple threads trying to run this function or update this value, correct?

Reply:
That's correct, there's only 1 thread writing to that value